File: client-android/lib/src/main/kotlin/stasis/client_android/lib/api/clients/caching/DatasetEntriesForDefinition.kt
package stasis.client_android.lib.api.clients.caching

import okio.ByteString
import stasis.client_android.lib.model.server.datasets.DatasetEntry
import stasis.client_android.lib.model.server.datasets.DatasetEntryId
import stasis.client_android.lib.utils.Try
import stasis.client_android.lib.utils.Try.Companion.map
import java.time.Instant
import java.util.UUID

data class DatasetEntriesForDefinition(
    val entries: Map<DatasetEntryId, Instant>,
    val latest: DatasetEntryId?
) {
    fun withEntry(entry: DatasetEntry): DatasetEntriesForDefinition {
        val updated = entries + (entry.id to entry.created)
        return copy(entries = updated, latest = updated.maxByOrNull { it.value }?.key)
    }

    companion object {
        operator fun invoke(entries: List<DatasetEntry>): DatasetEntriesForDefinition =
            DatasetEntriesForDefinition(
                entries = entries.associate { it.id to it.created },
                latest = entries.maxByOrNull { it.created }?.id
            )

        fun empty(): DatasetEntriesForDefinition =
            DatasetEntriesForDefinition(
                entries = emptyMap(),
                latest = null
            )

        fun DatasetEntriesForDefinition.toByteString(): ByteString =
            stasis.client_android.lib.model.proto.DatasetEntriesForDefinition(
                entries = this.entries.map { it.key.toString() to it.value.toEpochMilli() }.toMap(),
                latest = this.latest?.toString(),
            ).encodeByteString()

        fun ByteString.toDatasetEntriesForDefinition(): Try<DatasetEntriesForDefinition> = Try {
            stasis.client_android.lib.model.proto.DatasetEntriesForDefinition.ADAPTER.decode(this)
        }.map { entry ->
            DatasetEntriesForDefinition(
                entries = entry.entries.map { UUID.fromString(it.key) to Instant.ofEpochMilli(it.value) }.toMap(),
                latest = entry.latest?.let { UUID.fromString(it) }
            )
        }
    }
}

File: client-ios/lib/Sources/StasisClientLib/Api/Clients/Caching/DatasetEntriesForDefinition.swift
import Foundation

public struct DatasetEntriesForDefinition: Sendable, Equatable, Hashable {
    public let entries: [DatasetEntryId: Date]
    public let latest: DatasetEntryId?

    public init(entries: [DatasetEntryId: Date], latest: DatasetEntryId?) {
        self.entries = entries
        self.latest = latest
    }

    public init(entries: [DatasetEntry]) {
        let mapping = Dictionary(uniqueKeysWithValues: entries.map { ($0.id, $0.created) })
        self.entries = mapping
        self.latest = mapping.max(by: { $0.value < $1.value })?.key
    }

    public static func empty() -> DatasetEntriesForDefinition {
        DatasetEntriesForDefinition(entries: [:], latest: nil)
    }

    public func with(entry: DatasetEntry) -> DatasetEntriesForDefinition {
        var updated = entries
        updated[entry.id] = entry.created
        let newLatest = updated.max(by: { $0.value < $1.value })?.key
        return DatasetEntriesForDefinition(entries: updated, latest: newLatest)
    }
}
